added: exclusion violation custom error

manthan-jsharma opened the pull request:
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, open one before creating this pull request. -->
Fixes #2126 


<!-- Concisely describe what the pull request does. -->

**Technical details**
<!-- Add any other information or technical details about the implementation; or delete the section entirely. -->

**Screenshots**
<!-- Add screenshots to show the problem and the solution; or delete the section entirely. -->
<img width="1221" alt="Screenshot 2023-09-08 at 11 48 10 PM" src="https://github.com/centerofci/mathesar/assets/123374300/6bb09b08-542d-44f9-be02-cfb6942b1e3c">


## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. --> 
- [ ] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [ ] My pull request targets the `develop` branch of the repository
- [ ] My commit messages follow [best practices][best_practices].
- [ ] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

dmos62 commented on mathesar/api/serializers/records.py:
```diff
@@ -53,6 +53,12 @@ def update(self, instance, validated_data):
+            elif type(e.orig) == ExclusionViolation:
+                raise database_api_exceptions.ExclusionViolationAPIException(
+                    e,
+                    message="The requested update violates a exclusion constraint",
```
```suggestion
                    message="The requested update violates an exclusion constraint",
```

I suggest moving this message declaration into the default in the exception class. You did also edit that default there as well, but I'm not sure I understood the intent of that message. I think you can just replace it with this message. That way you'll be able to also get rid of the duplicate message in the other try/catch block you created in this file.

manthan-jsharma replied:
hi, @dmos62

> I suggest moving this message declaration into the default in the exception class 

As far as I understand, you're suggesting that we should store the message in a variable and then use that variable in both places so here is the commit link : https://github.com/centerofci/mathesar/pull/3200/commits/2c7888804afc104a0effd8ff17d366006afd6297

@dmos62, please let me know if I'm correct

dmos62 replied:
No. You now defined the default message for `ExclusionViolationAPIException` in the record serializer class. Instead, define it in the `ExclusionViolationAPIException` class. I.e. go to that class and give the `message` parameter a default value that corresponds to the variable that you created.

manthan-jsharma replied:
got it now,

here is the change : https://github.com/centerofci/mathesar/pull/3200/commits/42b7fb7ec542aa356e6a6cd77ef5d3d67a37903c

I've added a default value for message variable in ExclusionViolationAPIException

manthan-jsharma replied:
@dmos62, should I write test cases? 

If so, could you please advise me on which CSV file table to use for the test cases? I am having difficulty finding a table that allows me to reproduce an exclusion violation. 

Should I add a specific CSV file to reproduce it?

example 

| reservation_id  | room_number | check_in_date |  check_out_date |
| -------------- | --------------- | --------------- | ---------------- |
| 1                       | 101                     | 2023-09-10      | 2023-09-15       |

dmos62 reviewed (changes requested):
Approach looks good. Looking forward to seeing the tests.

rajatvijay commented:
@manthan-jsharma - python lint action is failing, please fix that too in addition to the review comments.

manthan-jsharma commented:
@rajatvijay, I ran the lint test locally and it passed. 

<img width="500" alt="Screenshot 2023-09-24 at 1 55 06 AM" src="https://github.com/centerofci/mathesar/assets/123374300/0be94ec0-71a4-4bfe-b366-fd76012f4b86">

\
&nbsp;


@rajatvijay, is it possible if this lint issue is coming after develop merge ? because I'm not using `listForRef` in the code for this PR

<img width="548" alt="Screenshot 2023-09-24 at 2 01 21 AM" src="https://github.com/centerofci/mathesar/assets/123374300/7867d4ea-df20-4029-b962-53808a30c45d">

<img width="800" alt="Screenshot 2023-09-24 at 2 01 50 AM" src="https://github.com/centerofci/mathesar/assets/123374300/fcb212cf-b036-4ac0-ab8c-ddf39fe12378">

Aritra8438 commented on mathesar/api/serializers/records.py:
```diff
@@ -80,6 +85,11 @@ def create(self, validated_data):
                     e,
                     status_code=status.HTTP_400_BAD_REQUEST
                 )
+            elif type(e.orig) == ExclusionViolation:
```
```suggestion
            elif type(e.orig) is ExclusionViolation:
```

manthan-jsharma replied:
thanks @Aritra8438 for pointing it out , i have pushed the changes pls mark as resolved

Aritra8438 commented on mathesar/api/serializers/records.py:
```diff
@@ -53,6 +53,11 @@ def update(self, instance, validated_data):
                     e,
                     status_code=status.HTTP_400_BAD_REQUEST
                 )
+            elif type(e.orig) == ExclusionViolation:
```
```suggestion
            elif type(e.orig) is ExclusionViolation:
```

manthan-jsharma replied:
thanks @Aritra8438 for pointing it out , i have pushed the changes pls mark as resolved

Aritra8438 reviewed (changes requested):
@manthan-jsharma, you were so close to discovering what was the error!
You might want to notice the line number where it's failing.

Though, it's weird that lint checks passed when you ran locally.
Hope this fixes the lint error.

Also, 
- if something is broken in the master branch, it will be evident from the repo test checks.
- If your pull request breaks after merging master branch into it, it should be fixed by you.

dmos62 reviewed (dismissed):
Looking better! Yes, you want a test for this. I wouldn't bother using a CSV file. Create the fixture you need via psycopg calls. Read other fixtures for inspiration.

seancolsen commented:
@manthan-jsharma Are you still planning to work on this?

manthan-jsharma commented:
Certainly, I will be completing it. Currently, I am in the process of understanding how to import CSV files for the test cases. It would be helpful if you could provide guidance on this matter. Apart from that, the work is nearly finished.

seancolsen commented:
@manthan-jsharma 

> It would be helpful if you could provide guidance on this matter

If you need help, please ask specific questions, taking care to explain exactly what problem you're having. Be clear about what steps you've taken, what expectation you have, and what observations you've made. If your questions are specific to the changes in this PR or are about implementation strategy, then the best place to ask is within this PR thread. However if your questions are more general questions about development processes and workflows, then it's probably better to ask in Matrix

manthan-jsharma commented:
Hello @seancolsen and @dmos62,


I have implemented the `custom message` for the exclusion violation, added `room_reservation.csv` for the test case, and made an effort to include the corresponding test case. However, I'm currently facing challenges in writing the class for the ExclusionViolation constraint, and this is the only remaining part.

The code for the test case has been pushed in this commit: [Link to the commit](https://github.com/centerofci/mathesar/pull/3200/commits/10428d01055afd25d62aa6fe884f1a8af75059ed).

@seancolsen and @dmos62, could you please guide me on how to write the expression for the exclusion constraint in this section of the code: [Link to the code](https://github.com/centerofci/mathesar/blob/10428d01055afd25d62aa6fe884f1a8af75059ed/db/constraints/base.py#L80)? Additionally, I am unsure if it is even possible to add the exclusion constraint using SQLAlchemy version 1.4.26. I couldn't find relevant information in the documentation; here is the link to the documentation: [SQLAlchemy Documentation](https://docs.sqlalchemy.org/en/14/).

seancolsen commented:
@manthan-jsharma, `@dmos62` will no longer be working on this ticket. I'm re-assigning this ticket to @Anish9901 who I hope will be able to answer your questions.

Anish9901 commented on mathesar/api/exceptions/database_exceptions/exceptions.py:
```diff
@@ -400,12 +400,40 @@ class ExclusionViolationAPIException(MathesarAPIException):
+                    "constraint": constraint.id,
+                    "constraint_columns": [c.id for c in constraint.columns],
+                }
+            except TypeError:
+                details = {
+                    "constraint": None,
+                }
```
When does this occur?

manthan-jsharma replied:
I've just handled an edge case. This code will only execute in the event of a failure within the try block.

Anish9901 replied:
Shouldn't that be handled using `Exception` as it a base class for all the python exceptions instead of `TypeError`?

Anish9901 commented on mathesar/api/exceptions/database_exceptions/exceptions.py:
```diff
@@ -400,12 +400,40 @@ class ExclusionViolationAPIException(MathesarAPIException):
+                }
+            )
+
+        exception_detail = get_default_exception_detail(
+            exception, self.error_code, message, field, details
+        )._asdict()
+        self.detail = [exception_detail]
+        self.status_code = status_code
```
use `super()` instead.

manthan-jsharma replied:
This is done, I've used super here

Anish9901 commented on mathesar/tests/api/test_record_api.py:
```diff
@@ -1464,3 +1464,33 @@ def test_record_patch_unique_violation(create_patents_table, client):
+    table = room_reservations_table(table_name)
+    id_column_id = table.get_column_name_id_bidirectional_map()['id']
+    columns_name_id_map = table.get_column_name_id_bidirectional_map()
+    # table.add_constraint(
+    #     ExclusionViolation(
+    #     what parameters should i pass ?
+    #     )
+    # )
```
You could just use raw SQL to add the required constraint for now.

manthan-jsharma replied:
I've attempted several SQL queries, but none  of them seems to be working successful. Additionally, I couldn't find any examples in the documentation on how to add exclusion constraints using SQL. Could you please provide guidance on adding exclusion constraints in SQL here @Anish9901

Anish9901 replied:
You are facing this problem because the data type of the columns of the CSV are getting inferred as `TEXT` instead of `DATE`

Anish9901 commented on mathesar/tests/api/test_record_api.py:
```diff
@@ -1464,3 +1464,33 @@ def test_record_patch_unique_violation(create_patents_table, client):
     assert actual_constraint_details['name'] == 'NASA unique record PATCH_pkey'
+
+
+def test_record_patch_exclusion_violation(room_reservations_table, client):
```
Please create a test case for `POST` as well.

Anish9901 reviewed (changes requested):
Hey @manthan-jsharma, I have requested some changes and addressed your queries in the code review please take a look at the specific comments.

manthan-jsharma commented on db/constraints/base.py:
```diff
@@ -64,3 +64,20 @@ def get_constraint_def_json(self):
         )
+
+
+class ExclusionViolation(Constraint):
```
I've removed the class

Anish9901 commented:
@manthan-jsharma I've added 12aa7e8 that takes care of the issues you were facing and the commit also include some general cleanups. Let me know if you have any concerns.
